File: bloque2.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.stattools import adfuller, kpss
import yfinance as yf
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# Función helper para prueba ADF
def adf_test(series, title=''):
    print(f'Augmented Dickey-Fuller Test: {title}')
    result = adfuller(series)
    print('ADF Statistic:', result[0])
    print('p-value:', result[1])
    print('Critical values:')
    for key, value in result[4].items():
        print(f'\t{key}: {value}')
    print('\n')

# Función helper para prueba KPSS
def kpss_test(series, title=''):
    print(f'KPSS Test: {title}')
    result = kpss(series)
    print('KPSS Statistic:', result[0])
    print('p-value:', result[1])
    print('Critical values:')
    for key, value in result[3].items():
        print(f'\t{key}: {value}')
    print('\n')

# Simulación de series temporales
np.random.seed(123)

# Ruido blanco (serie estacionaria)
norm_1 = np.random.normal(0, 1, 100)
ruido_1 = pd.Series(norm_1)
plt.figure(figsize=(10, 6))
plt.title('Ruido Blanco (σ = 1)')
plt.plot(ruido_1)
plt.grid(True)
plt.show()

# Ruido blanco con mayor varianza
norm_2 = np.random.normal(0, 3, 100)
ruido_2 = pd.Series(norm_2)
plt.figure(figsize=(10, 6))
plt.title('Ruido Blanco (σ = 3)')
plt.plot(ruido_2)
plt.grid(True)
plt.show()

# Caminata aleatoria sin deriva
ruido_blanco = np.concatenate(([0], np.random.normal(0, 1, 99)))
random_walk_sin_drift = np.cumsum(ruido_blanco)

plt.figure(figsize=(10, 6))
plt.title('Caminata Aleatoria sin Deriva')
plt.plot(random_walk_sin_drift)
plt.grid(True)
plt.show()

# Caminata aleatoria con deriva
deriva = np.repeat(0.4, 99)
ruido_blanco = np.random.normal(0, 1, 99)
random_walk_con_drift = np.cumsum(ruido_blanco + deriva)
random_walk_con_drift = np.concatenate(([0], random_walk_con_drift))

plt.figure(figsize=(10, 6))
plt.title('Caminata Aleatoria con Deriva')
plt.plot(random_walk_con_drift)
plt.grid(True)
plt.show()

# Pruebas de estacionariedad
print("Pruebas de estacionariedad para ruido blanco:")
adf_test(ruido_1, "Ruido Blanco σ=1")
adf_test(ruido_2, "Ruido Blanco σ=3")

print("Pruebas de estacionariedad para caminatas aleatorias:")
adf_test(random_walk_sin_drift, "Caminata Aleatoria sin Deriva")
adf_test(random_walk_con_drift, "Caminata Aleatoria con Deriva")

print("\nPruebas KPSS:")
kpss_test(ruido_1, "Ruido Blanco σ=1")
kpss_test(ruido_2, "Ruido Blanco σ=3")
kpss_test(random_walk_sin_drift, "Caminata Aleatoria sin Deriva")
kpss_test(random_walk_con_drift, "Caminata Aleatoria con Deriva")

# Para los datos de FRED, usaremos yfinance como alternativa
# Descargamos datos de ejemplo (S&P 500 como proxy)
sp500 = yf.download('^GSPC', start='2000-01-01', end=datetime.now())
# Verificamos las columnas disponibles
print("Columnas disponibles:", sp500.columns)

# Usamos 'Close' en lugar de 'Adj Close'
y1 = sp500['Close']
y2 = sp500['Volume']

# Log-transformación
log_y1 = 100 * np.log(y1)
log_y2 = 100 * np.log(y2)

# Visualización de series originales y transformadas
plt.figure(figsize=(12, 8))
plt.subplot(2,1,1)
plt.title('Serie Original (Precio de Cierre S&P 500)')
plt.plot(y1, label='Precio')
plt.legend()
plt.grid(True)
plt.subplot(2,1,2)
plt.title('Serie Log-transformada')
plt.plot(log_y1, label='Log-Precio', color='orange')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

# Pruebas de estacionariedad sobre niveles
print("\nPruebas sobre niveles:")
adf_test(y1, "Serie Y1 Original")
adf_test(log_y1, "Serie Y1 Log-transformada")
adf_test(y2, "Serie Y2 Original")
adf_test(log_y2, "Serie Y2 Log-transformada")

# Primeras diferencias
first_diff_y1 = np.diff(log_y1)
first_diff_y2 = np.diff(log_y2)

# Visualización de primeras diferencias
plt.figure(figsize=(10, 6))
plt.title('Primera Diferencia de Serie Log-transformada (Y1)')
plt.plot(first_diff_y1)
plt.grid(True)
plt.show()

# Pruebas ADF sobre primeras diferencias
print("\nPruebas sobre primeras diferencias:")
adf_test(first_diff_y1, "Primera Diferencia Y1")
adf_test(first_diff_y2, "Primera Diferencia Y2")

File: R scripts/Bloque_2.R

################################### CURSO DE ECONOMETRÍA DE SERIES DE TIEMPO 2025 ##################################################

#################### DEFINICION DE DIRECTORIO, CARGA DE ARCHIVOS Y PAQUETES ########################################################

## Establecemos el directorio de trabajo donde se encuentran nuestros archivos /

  # Verificamos que el directorio se haya definido correctamente

## Cargamos los paquetes necesarios para manipular datos y leer archivos Excel
library(tidyverse)
library(readxl)

## Importamos los datos desde un archivo Excel. Usamos "skip = 9" para omitir las primeras filas con títulos y notas.
inf <- read_excel("Anexo_Estadístico_del_Informe_Económico_07_05_2025 (1).xlsx", 
                  sheet = "CUADRO 9", skip = 9) 

inf <- inf[-1]  # Eliminamos la primera columna vacía (típico en Excel exportado desde PDFs u otras fuentes)
names(inf)[1] <- "mes"  # Renombramos la primera columna a "mes"
inf$mes <- seq(from = as.Date("1994-01-01"), by = "month", length.out = nrow(inf)) # Creamos una secuencia de fechas mensuales iniciando en enero de 1994, tantas como filas tenga el dataset
inf <- inf[1:(nrow(inf) - 3), ]  # Eliminamos las últimas 3 filas que pueden contener totales, notas o NA

## Cargamos datos de productividad y horas trabajadas desde la API de FRED (Federal Reserve Economic Data)
library(fredr)
library(zoo)  # Para series temporales indexadas por fecha

api_key <- ""  # Aquí debes ingresar tu clave personal de la API de FRED
fredr_set_key(api_key)

metadata_productivity <- fredr_series("OPHNFB")  # Metadatos de productividad no agrícola por hora trabajada

data_productivity <- fredr_series_observations("OPHNFB")  # Datos de la serie

data_hours <- fredr_series_observations("HOANBS")  # Horas trabajadas en el sector no agrícola

# Creamos objetos tipo zoo, útiles para trabajar con fechas
y1 <- zoo(data_productivity$value, order.by = as.Date(data_productivity$date))
y2 <- zoo(data_hours$value, order.by = as.Date(data_hours$date))

# Visualizamos ambas series
plot(y1)
plot(y2)

#################### PRUEBAS DE ESTACIONARIEDAD ########################################################

# Simulamos ruido blanco (serie estacionaria por definición)
norm_1 <- rnorm(100, mean = 0, sd = 1)
ruido_1 <- ts(norm_1, start = 1900, frequency = 1)
plot.ts(ruido_1)

# Ruido blanco con mayor varianza
norm_2 <- rnorm(100, mean = 0, sd = 3)
ruido_2 <- ts(norm_2, start = 1900, frequency = 1)
plot.ts(ruido_2)

# Simulamos una caminata aleatoria sin deriva (no estacionaria)
ruido_blanco <- c(0, rnorm(99, mean = 0, sd = 1))
caminata_aleatoria <- cumsum(ruido_blanco)  # Suma acumulada

df_sin_deriva <- data.frame(Valor_anterior = c(0, caminata_aleatoria[-100]),
                            Ruido_blanco = ruido_blanco,
                            Random_walk = caminata_aleatoria)

random_walk_sin_drift <- ts(df_sin_deriva$Random_walk , start = 1, frequency = 1)

# Ahora con deriva: cada paso incluye un incremento constante
deriva <- rep(0.4, 99)  # Constante positiva en cada paso
ruido_blanco <- rnorm(99, mean = 0, sd = 1)
caminata_aleatoria_con_deriva <- cumsum(ruido_blanco + deriva)

df_con_deriva <- data.frame(Valor_anterior = c(0, 0, caminata_aleatoria_con_deriva[-99]),
                            Ruido_blanco = c(0, ruido_blanco),
                            Deriva = c(0, deriva),
                            Random_Walk_with_Drift = c(0, caminata_aleatoria_con_deriva))

random_walk_con_drift <- ts(df_con_deriva$Random_Walk_with_Drift , start = 1, frequency = 1)

# Prueba ADF (Augmented Dickey-Fuller) para detectar raíces unitarias
# H0: No estacionaria | H1: Estacionaria
library(tseries)

adf.test(ruido_1)  # Esperamos rechazar H0 (es ruido blanco)
adf.test(ruido_2)

adf.test(random_walk_sin_drift)  # Esperamos no rechazar H0 (caminata aleatoria)
adf.test(random_walk_con_drift)

# Prueba KPSS (Kwiatkowski-Phillips-Schmidt-Shin)
# H0: Estacionaria | H1: No estacionaria
kpss.test(ruido_1)  # Esperamos no rechazar H0 (es estacionaria)
kpss.test(ruido_2)

kpss.test(random_walk_sin_drift)  # Esperamos rechazar H0 (no estacionaria)
kpss.test(random_walk_con_drift)

# Log-transformamos productividad y horas trabajadas para interpretar cambios como tasas de crecimiento
log_y1 <- 100 * log(y1)  # Multiplicamos por 100 para que los cambios se interpreten en %
log_y2 <- 100 * log(y2)

# Pruebas de estacionariedad sobre niveles
adf.test(y1)
adf.test(log_y1)
plot(y1)
plot(log_y1)

adf.test(y2)
adf.test(log_y2)

# Aplicamos primeras diferencias a las series logarítmicas para obtener tasas de crecimiento
first_diff_y1 <- diff(log_y1, differences = 1)
first_diff_y2 <- diff(log_y2, differences = 1)

plot(first_diff_y1)  # Estas deberían lucir más estacionarias visualmente
plot(first_diff_y2)

# Pruebas ADF sobre primeras diferencias: esperamos rechazar H0
adf.test(first_diff_y1)
adf.test(first_diff_y2)

print(first_diff_y1)  # Imprimimos para ver valores de las tasas de crecimiento









